### ISSUE TITLE
Node 15 and 16 do not export `_uv_dl*` symbols

### ISSUE THREAD

**Issue by bergundy** [Homebrew/homebrew-core#76893]
### `brew gist-logs <formula>` link OR `brew config` AND `brew doctor` output

```shell
$ brew gist-logs node
Error: No logs.
$ brew config
HOMEBREW_VERSION: 3.1.5
ORIGIN: https://github.com/Homebrew/brew
HEAD: 227445ece9414a82aa8929b91ca2e0cca28a561b
Last commit: 10 days ago
Core tap ORIGIN: https://github.com/Homebrew/homebrew-core
Core tap HEAD: b4e38c605e14bef7c2298e092deb8bdc1813394d
Core tap last commit: 68 minutes ago
Core tap branch: master
HOMEBREW_PREFIX: /opt/homebrew
HOMEBREW_CASK_OPTS: []
HOMEBREW_EDITOR: vim
HOMEBREW_MAKE_JOBS: 8
Homebrew Ruby: 2.6.3 => /System/Library/Frameworks/Ruby.framework/Versions/2.6/usr/bin/ruby
CPU: octa-core 64-bit arm_firestorm_icestorm
Clang: 12.0 build 1200
Git: 2.24.3 => /Library/Developer/CommandLineTools/usr/bin/git
Curl: 7.64.1 => /usr/bin/curl
macOS: 11.2.2-arm64
CLT: 12.4.0.0.1.1610135815
Xcode: N/A
Rosetta 2: false
$ brew doctor
Your system is ready to brew.
```

### 

- [X] I ran `brew update` and am still able to reproduce my issue.
- [X] I have resolved all warnings from `brew doctor` and that did not fix my problem.

### What were you trying to do (and why)?

Run a NodeJS package which requires the `_uv_dlopen` symbol to be exported.

### What happened (include all command output)?

NodeJS runs fine until the package [isolated-vm](https://github.com/laverdet/isolated-vm) tries to load a binary module using `_uv_dlopen`.

Getting empty output when searching for `_uv_dl` in the binary's symbol table.
```shell
$ which node
/opt/homebrew/bin/node
$ node --version
v16.1.0
/opt/homebrew/bin/node
$ nm -g $(which node) | grep _uv_dl
```

### What did you expect to happen?

I expect to see the symbol exported as I see with node installed with `nvm`.
```
$ nvm install 16
Downloading and installing node v16.1.0...
Local cache found: ${NVM_DIR}/.cache/bin/node-v16.1.0-darwin-arm64/node-v16.1.0-darwin-arm64.tar.xz
Checksums match! Using existing downloaded archive ${NVM_DIR}/.cache/bin/node-v16.1.0-darwin-arm64/node-v16.1.0-darwin-arm64.tar.xz
Now using node v16.1.0 (npm v7.11.2)
$ nm -g $(which node) | grep _uv_dl
0000000100973db0 T _uv_dlclose
0000000100973e50 T _uv_dlerror
0000000100973d30 T _uv_dlopen
0000000100973de8 T _uv_dlsym
```

### Step-by-step reproduction instructions (by running `brew` commands)

```shell
brew install node
nm -g $(which node) | grep _uv_dl  # Empty output
```

**Comment by SMillerDev** [comment]
Is this something that should be present in a default node install? And what provides this?

**Comment by bergundy** [comment]
It was present in the node installations up until node@15 in brew, reexported from libuv.

I  tried to run the `configure` command from the [formula](https://github.com/Homebrew/homebrew-core/blob/master/Formula/node.rb) to reproduce but I'm not sure how to get the output of `Formula["foo"].include` in my shell.

**Comment by chrmoritz** [comment]
I assume, that might be a side effect of https://github.com/Homebrew/homebrew-core/pull/74698.

**Comment by carlocab** [comment]
> I assume, that might be a side effect of #74698.

Yep, seems likely.

> I tried to run the `configure` command from the [formula](https://github.com/Homebrew/homebrew-core/blob/master/Formula/node.rb) to reproduce but I'm not sure how to get the output of `Formula["foo"].include` in my shell.

`Formula["foo"].include` is just `$(brew --prefix foo)/include`.

**Comment by carlocab** [comment]
Is `node` supposed to re-export those when `libuv` is dynamically linked rather than statically linked? Because it doesn't do so for OpenSSL.

```
❯ nm -g ~/.nvm/versions/node/v16.1.0/bin/node | rg -q '\W_tls' && echo "exports _tls* symbols"
exports _tls* symbols
❯ nm -g $(brew --prefix node)/bin/node | rg -q '\W_tls' && echo "exports _tls* symbols"
❯ echo $?
1
```

**Comment by chrmoritz** [comment]
> Is `node` supposed to re-export those when `libuv` is dynamically linked rather than statically linked? Because it doesn't do so for OpenSSL.

Yeah. From https://nodejs.org/api/addons.html :
> * Node.js includes other statically linked libraries including OpenSSL. These other libraries are located in the `deps/` directory in the Node.js source tree. **Only the libuv, OpenSSL, V8 and zlib symbols are purposefully re-exported by Node.js** and may be used to various extents by addons. See [Linking to libraries included with Node.js](https://nodejs.org/api/addons.html#addons_linking_to_libraries_included_with_node_js) for additional information.

**Comment by carlocab** [comment]
V8 is statically linked into Node.

The things you quote from nodejs.org describe Node's behaviour when the libraries are statically linked; I'm not sure it implies anything about what it should be doing when the libraries are linked dynamically. Does Node export these symbols when compiled to link against a shared OpenSSL or libuv, without using Homebrew to build Node?

For reference, Debian and Arch build Node with dynamic linkage to these libraries.

**Comment by carlocab** [comment]
Fired up a Docker container to install Node 16 from the [NodeSource](https://github.com/nodesource/distributions) repository. Not exporting these symbols seems entirely expected behaviour to me:
```
root@f6fdd1bea495:/home/linuxbrew# nodejs --version
v16.1.0
root@f6fdd1bea495:/home/linuxbrew# nm -g $(which nodejs) | grep '_uv_dl'; echo $?
1
```
Unless there's some reason these symbols need to be exported on macOS but not on Linux. (Or if `nm -g` actually does something different on Linux that I'm not aware of.)

**Comment by chrmoritz** [comment]
> Does Node export these symbols when compiled to link against a shared OpenSSL or libuv, without using Homebrew to build Node?

It doesn't look so. On the other hand breaking some native addons when dynamically linking deps seems to be not the expected behavior either.

Related issue: https://github.com/laverdet/isolated-vm/issues/238

**Comment by carlocab** [comment]
> On the other hand breaking some native addons when dynamically linking deps seems to be not the expected behavior either.

This is unfortunate, but I'm not sure there's much we can or should do about it. We could perhaps statically link libuv, but statically linking OpenSSL is, IMHO, out of the question. Linking OpenSSL dynamically will also produce the same problems when native addons start looking for `_tls*` symbols in the `node` binary.

This is probably why projects ship `pc` files or `foo-config` executables to simplify linking, but Node doesn't do this.

**Comment by bergundy** [comment]
Looks like the cause is linker optimization stripping out functions which are not used AFAICT by looking at [node.gypi](https://github.com/nodejs/node/blob/7c8a60851c459ea18afbfc54bfc8cf7394ea56c3/node.gypi#L2).
I'm not sure what the correct behavior is, either isolated-vm should avoid using `_uv_dlopen` or NodeJS should always export the `_uv_dl*` functions.
As a temporary solution, the issue should be fixed by brew statically linking libuv.

**Comment by carlocab** [comment]
> As a temporary solution, the issue should be fixed by brew statically linking libuv.

That's not a good temporary solution, I'm afraid. It pretty much guarantees that it won't be temporary. We should statically link libuv only if we think it's the long-term solution here.

**Comment by chrmoritz** [comment]
https://github.com/nodejs/node/blob/master/node.gypi#L2-L8:
>  \# Linker optimizes out functions that are not used. When force_load=true,
>  \# --whole-archive,force_load and /WHOLEARCHIVE are used to include
>  \# all obj files in static libs into the executable or shared lib.

Does `superenv` do something, which would cause the `--whole-archive` to be filtered out?

**Comment by Bo98** [comment]
It shouldn't be, no. Can fairly easily confirm with the `.cc` logs the superenv creates.

**Comment by bergundy** [comment]
Now that I think of it, why does brew need to compile node@16?
From the [release notes](https://medium.com/the-node-js-collection/node-js-16-available-now-7f5099a97e70):

> Node.js v16.0.0 will be the first release where we ship prebuilt binaries for Apple Silicon. While we’ll be providing separate tarballs for the Intel (darwin-x64) and ARM (darwin-arm64) architectures the macOS installer (.pkg) will be shipped as a ‘fat’ (multi-architecture) binary.

**Comment by SMillerDev** [comment]
> Now that I think of it, why does brew need to compile node@16?

Because brew is an open source package manager and we compile that source in CI (except the "cask" packages)

**Comment by carlocab** [comment]
If someone wants to test this, I think this might help:
```diff
diff --git a/Formula/node.rb b/Formula/node.rb
index 16918e36c0..c3c7a6bf8b 100644
--- a/Formula/node.rb
+++ b/Formula/node.rb
@@ -65,6 +65,7 @@ class Node < Formula
     ]
     args << "--tag=head" if build.head?
 
+    on_macos { ENV.append "LDFLAGS", "-Wl,-force_load,#{Formula["libuv"].opt_lib}/libuv.a" }
     system "./configure", *args
     system "make", "install"

```
That said, I suspect this is really just a workaround, and this should probably be reported upstream instead/as well.

**Comment by bergundy** [comment]
I'll open an issue upstream.

**Comment by fsaari1985** [comment]
2 errors generated.
make[1]: *** [/private/tmp/node-20210726-15402-f88etf/node-v16.5.0/out/Release/obj.target/v8_libplatform/deps/v8/src/libplatform/default-platform.o] Error 1
make[1]: *** Waiting for unfinished jobs....
rm 04b6ed1aaec2a380303542cc2608ffc7870f1bea.intermediate
make: *** [node] Error 2

**Comment by fsaari1985** [comment]
anyone have a solution for this?

**Comment by bergundy** [comment]
FTR, this issue is solved in node 16.6 (and perhaps 16.5 but I didn't test it).
https://github.com/nodejs/node/issues/39210
